contracts/upgradeable_contracts/arbitrary_message/BasicAMB.sol

```diff
@@ -0,0 +1,81 @@
+pragma solidity 0.4.24;
+
+import "../../upgradeability/EternalStorage.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/AddressUtils.sol";
+import "../Validatable.sol";
+import "../BasicBridge.sol";
+
+contract BasicAMB is BasicBridge {
+    bytes internal constant SUBSIDIZED_MODE = bytes(abi.encodePacked("AMB-subsidized-mode"));
+    bytes internal constant DEFRAYAL_MODE = bytes(abi.encodePacked("AMB-defrayal-mode"));
+    bytes32 internal constant HOME_TO_FOREIGN_MODE = keccak256(abi.encodePacked("homeToForeignMode"));
```

Comment:
Is it really needed that the corresponding contract knows its own mode? My understanding that it is more important to know the mode of the opposite site to manage proper set of arguments to relay.

So, my suggestion is to have move functionality related to the mode of Home side to `BasicForeignAMB` directly and the mode of the Foreign side to `BasicHomeAMB`. The method `initialize` of `BasicAMB` will be internal and `Basic...AMB` contracts will call it to perform initialization for common parameters.

Here is an example to illustrate what I mean:
```solidity
contract BasicAMB is BasicBridge {
    bytes32 internal constant SUBSIDIZED_MODE = keccak256(abi.encodePacked("AMB-subsidized-mode"));
    bytes32 internal constant DEFRAYAL_MODE = keccak256(abi.encodePacked("AMB-defrayal-mode"));
    bytes32 internal constant MAX_GAS_PER_TX = keccak256(abi.encodePacked("maxGasPerTx"));

    function _initialize(
        address _validatorContract,
        uint256 _maxGasPerTx,
        uint256 _gasPrice,
        uint256 _requiredBlockConfirmations,
        address _owner
    ) internal {
    . . .
    }

    function getBridgeMode() external pure returns (bytes4 _data) {
        return bytes4(keccak256(abi.encodePacked("arbitrary-message-bridge-core")));
    }

    function maxGasPerTx() public view returns (uint256) {
        return uintStorage[MAX_GAS_PER_TX];
    }

     function setMaxGasPerTx(uint256 _maxGasPerTx) external onlyOwner {
        require(_maxGasPerTx > 0);
        uintStorage[MAX_GAS_PER_TX] = _maxGasPerTx;
    }

contract BasicForeignAMB is BasicAMB, MessageDelivery, MessageProcessor {
    bytes32 internal constant FOREIGN_TO_HOME_MODE = keccak256(abi.encodePacked("foreignToHomeMode"));

    function initialize(
        address _validatorContract,
        uint256 _maxGasPerTx,
        uint256 _gasPrice,
        uint256 _requiredBlockConfirmations,
        address _owner
    ) public returns (bool) {
        require(!isInitialized());
        _initialize(_validatorContract, _maxGasPerTx, _gasPrice, _requiredBlockConfirmations, _owner);

        bytesStorage[FOREIGN_TO_HOME_MODE] = SUBSIDIZED_MODE;
        setInitialize();

        emit ForeignToHomeModeChanged(SUBSIDIZED_MODE);

        return isInitialized();
    }

     function setSubsidizedModeForForeignToHome() external onlyOwner {
        bytes32Storage[FOREIGN_TO_HOME_MODE] = SUBSIDIZED_MODE;
        emit ForeignToHomeModeChanged(SUBSIDIZED_MODE);
    }

     function setDefrayalModeForForeignToHome() external onlyOwner {
        bytes32Storage[FOREIGN_TO_HOME_MODE] = DEFRAYAL_MODE;
        emit ForeignToHomeModeChanged(DEFRAYAL_MODE);
    }

     function foreignToHomeMode() public view returns (bytes) {
        return bytes32Storage[FOREIGN_TO_HOME_MODE];
    }
. . .
. . .
}
```

Note, that I changed the data type for constants `SUBSIDIZED_MODE` and `DEFRAYAL_MODE` to reduce the gas usage. 
I also added event to notify about the mode change and make the mode by default `SUBSIDIZED_MODE` since the defrayal mode is not ready yet.

By moving the corresponding method to handle the mode to specific contract we also reduce gas usage on the deployment stage.

Reply:
My understanding is that messages that are transferred from home to be executed on foreign could be set to one mode, and messages that are trasferred from foreign to be executed on Home could be set to the same or another mode. Is this correct?

If this is true, for example on Home side, the contract needs to know how to pass the message to foreign when `requireToPassMessage` is called. So in this case it needs to know the mode of the contract on other network to know how will be the message processed. This is done on `MessageDelivery`.

https://github.com/poanetwork/poa-bridge-contracts/blob/c929d2c4083852ff66967164983818495226923f/contracts/upgradeable_contracts/arbitrary_message/MessageDelivery.sol#L8-L12

On another example, Home receives a message to process from foreign. In this case it needs to know its own mode to be able to run the proper checks. This is done on `MessageProcessor` 

https://github.com/poanetwork/poa-bridge-contracts/blob/c929d2c4083852ff66967164983818495226923f/contracts/upgradeable_contracts/arbitrary_message/MessageProcessor.sol#L36-L48

Definitions from Home contract:

https://github.com/poanetwork/poa-bridge-contracts/blob/c929d2c4083852ff66967164983818495226923f/contracts/upgradeable_contracts/arbitrary_message/HomeAMB.sol#L13-L23

So at least the way is implemented now it looks like the contract needs to know its own mode too, unless I'm missing something or some way to avoid this.

>Note, that I changed the data type for constants SUBSIDIZED_MODE and DEFRAYAL_MODE to reduce the gas usage.
I also added event to notify about the mode change and make the mode by default SUBSIDIZED_MODE since the defrayal mode is not ready yet.

Thanks, I'll apply these changes

Reply:
>Note, that I changed the data type for constants SUBSIDIZED_MODE and DEFRAYAL_MODE to reduce the gas usage.

Looking at the code, our `EternalStorage` does not provide a`bytes32Storage` mapping. To avoid modifying the storage do you think we can use `uint256` instead of `bytes32` so we can store the information on `uintStorage` and also save gas compairing with the previous `bytes` definition? In that case the variables could be defined as 
```solidity
    uint256 internal constant SUBSIDIZED_MODE = 0;
    uint256 internal constant DEFRAYAL_MODE = 1;
```

Reply:
> In this case it needs to know its own mode to be able to run the proper checks.

Thanks for clarification I was not aware about these checks.

> we can store the information on uintStorage and also save gas compairing with the previous bytes definition

yes, let's use their integer representation then.

Reply:
Applied changes https://github.com/poanetwork/poa-bridge-contracts/pull/77/commits/e7cf4fa8ab20a8a1b3f8b92880caeeb6f5dbee9a